pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -339,6 +340,15 @@ func newQueryMiddlewares(
 		newStepAlignMiddleware(limits, log, registerer),
 	)
 
+	if cfg.CacheResults && cfg.CacheErrors {
+		// TODO: Use a real TTL
+		queryRangeMiddleware = append(
```

Comment:
an additional middleware increases the cache calls - now we look up one more key from the cache sequentially. I'm mostly thinking about increasing cache lookup RPCs and increasing latency because we do this cache lookup before we proceed with the request. Did you consider including this logic in the current split&cache middleware?

Reply:
I considered it. I wanted to avoid changes to the split&cache middleware because it's quite complicated and very focused on caching results. Adding caching of errors to it would add more complexity. 

I also wanted to keep the error caching _before_ queries are split so that we have the best chance of avoiding work when any part of a query is going to fail. We would have do a lot more work if each individual split or sharded query can fail.

I'm not particularly worried about the latency added here for each request because it's only a single cache lookup which is a few milliseconds on hit or miss.

Reply:
ok, I guess if we measure these latencies it will be ok and we'd know if it's too much. Maybe we can stack them in the cache lookup panel in the reads dashboard?
<img width="998" alt="Screenshot 2024-08-20 at 17 20 21" src="https://github.com/user-attachments/assets/e70142bf-cf54-4fa8-9a2f-25c0a1f3e94e">

Reply:
Testing in a dev cell, this middleware ended up taking about 500us on cache hit or miss. I'll follow up this PR with some changes to the reads/queries dashboards if this turns out to be useful.